Source/JavaScriptCore/runtime/SetPrototype.cpp

```diff
@@ -175,6 +181,170 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncSize, (JSGlobalObject* globalObject, CallFr
+    }
+    return JSValue::encode(result);
+}
+
+JSC_DEFINE_HOST_FUNCTION(setProtoFuncIntersection, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSSet* thisSet = getSet(globalObject, callFrame->thisValue());
+    RETURN_IF_EXCEPTION(scope, { });
+
+    JSValue otherValue = callFrame->argument(0);
+
+    if (otherValue.isObject() && setPrimordialWatchpointIsValid(vm, asObject(otherValue))) [[likely]] {
+        if (auto* otherSet = jsDynamicCast<JSSet*>(otherValue.asCell())) [[likely]] {
+            scope.release();
+            return fastSetIntersection(globalObject, thisSet, otherSet);
+        }
+    }
+
+    uint32_t size = getSetSizeAsInt(globalObject, otherValue);
+    RETURN_IF_EXCEPTION(scope, { });
+
+    ASSERT(otherValue.isObject());
+    JSObject* otherObject = asObject(otherValue);
+
+    JSValue has = otherObject->get(globalObject, vm.propertyNames->has);
+    RETURN_IF_EXCEPTION(scope, { });
+    if (!has.isCallable()) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "Set.prototype.intersection expects other.has to be callable");
+
+    JSValue keys = otherObject->get(globalObject, vm.propertyNames->keys);
+    RETURN_IF_EXCEPTION(scope, { });
+    if (!keys.isCallable()) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "Set.prototype.intersection expects other.keys to be callable");
+
+    JSSet* result = JSSet::create(vm, globalObject->setStructure());
+    if (thisSet->size() <= size) {
+        JSCell* storageCell = thisSet->storageOrSentinel(vm);
+        if (storageCell == vm.orderedHashTableSentinel())
+            return JSValue::encode(result);
+
+        JSSet::Storage& storage = *jsCast<JSSet::Storage*>(storageCell);
+        JSSet::Helper::Entry entry = 0;
+        CallData hasCallData = JSC::getCallData(has);
+        while (true) {
+            storageCell = JSSet::Helper::nextAndUpdateIterationEntry(vm, storage, entry);
+            if (storageCell == vm.orderedHashTableSentinel())
+                break;
+
+            storage = *jsCast<JSSet::Storage*>(storageCell);
+            entry = JSSet::Helper::iterationEntry(storage) + 1;
+            JSValue entryKey = JSSet::Helper::getIterationEntryKey(storage);
+
+            JSValue hasResult;
+            if (hasCallData.type == CallData::Type::JS) [[likely]] {
+                CachedCall cachedCall(globalObject, jsCast<JSFunction*>(has), 1);
+                RETURN_IF_EXCEPTION(scope, { });
+                hasResult = cachedCall.callWithArguments(globalObject, otherValue, entryKey);
+                RETURN_IF_EXCEPTION(scope, { });
+            } else {
+                MarkedArgumentBuffer args;
+                args.append(entryKey);
+                ASSERT(!args.hasOverflowed());
+                hasResult = call(globalObject, has, hasCallData, otherValue, args);
+                RETURN_IF_EXCEPTION(scope, { });
+            }
```

Comment:
`CachedCall` is faster when we set up this outside of the loop.

Reply:
Fixed!